### PR TITLE
1084 dont allow selection of private readable repository mode

### LINKED ISSUE #1084: Don't allow selection of private_readable repository mode
Also 
* rename 'public readable' to 'public readable, private writable' *only* in the view.
* put a link to the wiki page near the drop down box to communicate that happens with the permissions.

### PR DESCRIPTION
**Pull request by eugenk**
This shall fix #1084. By splitting the access options into two hashes, we can disable `private_rw` for non-mirrors.

### CODE REVIEW

**Comment by SGraef** [comment]
I think it is not correct that we can't make a writable fork or is this intendet behaviour?

**Comment by SGraef** [comment]
we should not allow to select a mode which we don't support for the mirror-repositorytype so the mode Publicy readible, private writetabe should be remove from the list when choosing a mirror

**Comment by eugenk** [comment]
Todo:
* [] Fix #1463 here

**Comment by eugenk** [comment]
I decided to not implement 1463 here. Depending on the answer there. it may change too much for this pull request.

**Comment by eugenk** [comment]
The select box should be fixed now for forks.

File: spec/models/ability_spec.rb [hunk 1]
```diff
@@ -192,29 +192,39 @@
...
+      it 'reader, editor should not be allowed' do
         [reader, editor].each do |role|
-          Ability.new(role, nil).should_not be_able_to(:update, item)
+          expect(Ability.new(role, nil)).to_not be_able_to(:update, item)
         end
       end
 
-      it 'owner should not be allowed' do
-        Ability.new(owner, nil).should be_able_to(:update, item)
+      it 'owner should be allowed' do
```
Comment (eugenk) [inline spec/models/ability_spec.rb]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: spec/models/ability_spec.rb [hunk 2]
```diff
@@ -192,29 +192,39 @@
...
-          Ability.new(role, nil).should be_able_to(:show, item)
+        [editor, owner].each do |role|
+          expect(Ability.new(role, nil)).to be_able_to(:show, item)
         end
       end
     end
 
     context 'update:' do
-      it 'reader, editor should be allowed' do
+      it 'reader, editor should not be allowed' do
```
Comment (eugenk) [inline spec/models/ability_spec.rb]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: spec/models/ability_spec.rb [hunk 3]
```diff
@@ -192,29 +192,39 @@
...
+        expect(Ability.new(reader, nil)).to_not be_able_to(:write, item)
+      end
+
+      it 'be allowed: to read' do
+        expect(Ability.new(reader, nil)).to be_able_to(:show, item)
+      end
+    end
+
+    context 'editor, owner' do
+      it 'be allowed: to write' do
```
Comment (eugenk) [inline spec/models/ability_spec.rb]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: spec/models/ability_spec.rb [hunk 4]
```diff
@@ -192,29 +192,39 @@
...
-          Ability.new(role, nil).should_not be_able_to(:write, item)
+        expect(Ability.new(reader, nil)).to_not be_able_to(:write, item)
+      end
+
+      it 'be allowed: to read' do
+        expect(Ability.new(reader, nil)).to be_able_to(:show, item)
+      end
+    end
+
+    context 'editor, owner' do
```
Comment (eugenk) [inline spec/models/ability_spec.rb]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: spec/models/ability_spec.rb [hunk 5]
```diff
@@ -192,29 +192,39 @@
...
 
-    context 'reader, editor, owner' do
+    context 'reader' do
       it 'not be allowed: to write' do
-        [reader, editor, owner].each do |role|
-          Ability.new(role, nil).should_not be_able_to(:write, item)
+        expect(Ability.new(reader, nil)).to_not be_able_to(:write, item)
+      end
+
+      it 'be allowed: to read' do
```
Comment (eugenk) [inline spec/models/ability_spec.rb]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: spec/models/ability_spec.rb [hunk 6]
```diff
@@ -192,29 +192,39 @@
       end
     end
 
-    context 'reader, editor, owner' do
+    context 'reader' do
```
Comment (eugenk) [inline spec/models/ability_spec.rb]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: spec/models/ability_spec.rb [hunk 7]
```diff
@@ -172,10 +172,10 @@
...
   end
 
-  context 'Private read-only Repository' do
-    let(:editor){ create :user } # editor
-    let(:reader){ create :user } # reader
-    let(:item){   create(:repository, access: 'private_r', user: owner) }
+  context 'Private Repository' do
+    let(:editor) { create :user } # editor
+    let(:reader) { create :user } # reader
+    let(:item) { create(:repository, access: 'private_rw', user: owner) }
```
Comment (eugenk) [inline spec/models/ability_spec.rb]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: spec/models/ability_spec.rb [hunk 8]
```diff
@@ -172,10 +172,10 @@
     end
   end
 
-  context 'Private read-only Repository' do
-    let(:editor){ create :user } # editor
-    let(:reader){ create :user } # reader
-    let(:item){   create(:repository, access: 'private_r', user: owner) }
+  context 'Private Repository' do
```
Comment (eugenk) [inline spec/models/ability_spec.rb]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: app/views/repositories/_form.html.haml [hunk 9]
```diff
@@ -16,4 +15,9 @@
         = check_box_tag :un_mirror
         = t 'repository.un_mirror.text'
 
+  = f.input :access, collection: resource.mirror? ? access_options_mirror : access_options, include_blank: false, hint: access_change_hint
```
Comment (eugenk) [inline app/views/repositories/_form.html.haml]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: app/views/repositories/_form.html.haml [hunk 10]
```diff
@@ -16,4 +15,9 @@
...
+  .access-options.hide
+    = select_tag :access_options_non_mirror, options_for_select(access_options)
+    = select_tag :access_options_mirror, options_for_select(access_options_mirror)
```
Comment (eugenk) [inline app/views/repositories/_form.html.haml]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: app/models/repository/validations.rb [hunk 11]
```diff
@@ -118,9 +113,10 @@ def validate(record)
 
   class SourceTypeValidator < ActiveModel::Validator
     def validate(record)
-      if record.mirror? && !record.source_type.present?
+      if record.remote? && !record.source_type.present?
         record.errors[:source_address] = 'not a valid remote repository '\
-          "or not accessible (types supported: #{SOURCE_TYPES.join(', ')})"
+          'or not accessible '\
```
Comment (eugenk) [inline app/models/repository/validations.rb]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: app/models/repository/validations.rb [hunk 12]
```diff
@@ -24,19 +24,14 @@ module Repository::Validations
...
-      if record.mirror? && (record.private_rw? || record.public_rw?)
-        record.errors[:access] =
-          'Error! Write access is not allowed for a mirrored repositry.'
-      end
-    end
+      inclusion: {in: Repository::Access::OPTIONS},
+      unless: :mirror?
+    validates :access,
+      presence: true,
+      inclusion: {in: Repository::Access::OPTIONS_MIRROR},
```
Comment (eugenk) [inline app/models/repository/validations.rb]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: app/models/repository/validations.rb [hunk 13]
```diff
@@ -24,19 +24,14 @@ module Repository::Validations
...
-  end
-
-  class RemoteAccessValidator < ActiveModel::Validator
-    def validate(record)
-      if record.mirror? && (record.private_rw? || record.public_rw?)
-        record.errors[:access] =
-          'Error! Write access is not allowed for a mirrored repositry.'
-      end
-    end
+      inclusion: {in: Repository::Access::OPTIONS},
```
Comment (eugenk) [inline app/models/repository/validations.rb]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: app/helpers/repositories_helper.rb [hunk 14]
```diff
@@ -34,13 +34,11 @@ def access_change_hint
...
-        k.to_s.split('_')[1] == 'r'
-      else
-        true
-      end
-    end.invert
+    t('repository.access.options').invert
+  end
+
+  def access_options_mirror
+    t('repository.access.options_mirror').invert
```
Comment (eugenk) [inline app/helpers/repositories_helper.rb]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: app/models/repository/validations.rb [hunk 15]
```diff
@@ -24,19 +24,14 @@ module Repository::Validations
...
+      unless: :mirror?
+    validates :access,
+      presence: true,
```
Comment (eugenk) [inline app/models/repository/validations.rb]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

File: app/helpers/repositories_helper.rb [hunk 16]
```diff
@@ -34,13 +34,11 @@ def access_change_hint
...
 
   def access_options
-    t('repository.access.options').select do |k,v|
-      if @repository.mirror?
-        k.to_s.split('_')[1] == 'r'
-      else
-        true
-      end
-    end.invert
+    t('repository.access.options').invert
```
Comment (eugenk) [inline app/helpers/repositories_helper.rb]: No thanks. You should stop suggesting this as soon as #1574 gets merged.

**Comment by SGraef** [comment]
when i start to begin write an source address and select mirror and then remove the sourceaddress i still have only the possibilities of an mirror repo

**Comment by eugenk** [comment]
Good catch. Fixed with a4ab43b.

**Comment by SGraef** [comment]
:+1:

**Pull request merged**